src/content_mapblock.cpp

```diff
@@ -525,1633 +371,955 @@ static TileSpec getSpecialTile(const ContentFeatures &f,
-				aabb3f( g,-g,-g, g, g, g), // x+
-				aabb3f(-g,-g,-g,-g, g, g), // x-
-				aabb3f(-g,-g, g, g, g, g), // z+
-				aabb3f(-g,-g,-g, g, g,-g)  // z-
-			};
+	if (flowing) {
+		// Default downwards-flowing texture animation goes from
+		// -Z towards +Z, thus the direction is +Z.
+		// Rotate texture to make animation go in flow direction
+		// Positive if liquid moves towards +Z
+		f32 dz = (corner_levels[0][0] + corner_levels[0][1]) -
+		         (corner_levels[1][0] + corner_levels[1][1]);
+		// Positive if liquid moves towards +X
+		f32 dx = (corner_levels[0][0] + corner_levels[1][0]) -
+		         (corner_levels[0][1] + corner_levels[1][1]);
+		f32 tcoord_angle = atan2(dz, dx) * core::RADTODEG;
+		v2f tcoord_center(0.5, 0.5);
+		v2f tcoord_translate(blockpos_nodes.Z + p.Z, blockpos_nodes.X + p.X);
+		tcoord_translate.rotateBy(tcoord_angle);
+		tcoord_translate.X -= floor(tcoord_translate.X);
+		tcoord_translate.Y -= floor(tcoord_translate.Y);
+
+		for (int i = 0; i < 4; i++) {
+			vertices[i].TCoords.rotateBy(tcoord_angle, tcoord_center);
+			vertices[i].TCoords += tcoord_translate;
+		}
 
-			// table of node visible faces, 0 = invisible
-			int visible_faces[6] = {0,0,0,0,0,0};
-
-			// table of neighbours, 1 = same type, checked with g_26dirs
-			int nb[18] = {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
-
-			// g_26dirs to check when only horizontal merge is allowed
-			int nb_H_dirs[8] = {0,2,3,5,10,11,12,13};
-
-			content_t current = n.getContent();
-			content_t n2c;
-			MapNode n2;
-			v3s16 n2p;
-
-			// neighbours checks for frames visibility
-
-			if (!H_merge && V_merge) {
-				n2p = blockpos_nodes + p + g_26dirs[1];
-				n2 = data->m_vmanip.getNodeNoEx(n2p);
-				n2c = n2.getContent();
-				if (n2c == current || n2c == CONTENT_IGNORE)
-					nb[1] = 1;
-				n2p = blockpos_nodes + p + g_26dirs[4];
-				n2 = data->m_vmanip.getNodeNoEx(n2p);
-				n2c = n2.getContent();
-				if (n2c == current || n2c == CONTENT_IGNORE)
-					nb[4] = 1;
-			} else if (H_merge && !V_merge) {
-				for(i = 0; i < 8; i++) {
-					n2p = blockpos_nodes + p + g_26dirs[nb_H_dirs[i]];
-					n2 = data->m_vmanip.getNodeNoEx(n2p);
-					n2c = n2.getContent();
-					if (n2c == current || n2c == CONTENT_IGNORE)
-						nb[nb_H_dirs[i]] = 1;
-				}
-			} else if (H_merge && V_merge) {
-				for(i = 0; i < 18; i++)	{
-					n2p = blockpos_nodes + p + g_26dirs[i];
-					n2 = data->m_vmanip.getNodeNoEx(n2p);
-					n2c = n2.getContent();
-					if (n2c == current || n2c == CONTENT_IGNORE)
-						nb[i] = 1;
-				}
-			}
+		std::swap(vertices[0].TCoords, vertices[2].TCoords);
+	}
 
-			// faces visibility checks
-
-			if (!V_merge) {
-				visible_faces[0] = 1;
-				visible_faces[1] = 1;
-			} else {
-				for(i = 0; i < 2; i++) {
-					n2p = blockpos_nodes + p + dirs[i];
-					n2 = data->m_vmanip.getNodeNoEx(n2p);
-					n2c = n2.getContent();
-					if (n2c != current)
-						visible_faces[i] = 1;
-				}
-			}
+	static const u16 indices[] = { 0, 1, 2, 2, 3, 0 };
```

Comment:
btw, `g_6dirs`, `g_26dirs` and `g_27dirs` all contain the same data, just cut at different point, so all could be replaced with `g_27dirs`